source/slang/slang-ir-com-interface.cpp

```diff
@@ -0,0 +1,98 @@
+// slang-ir-com-interface.cpp
+#include "slang-ir-com-interface.h"
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+
+struct ComInterfaceLoweringContext
+{
+    IRModule* module;
+    DiagnosticSink* diagnosticSink;
+
+    SharedIRBuilder sharedBuilder;
+
+    Dictionary<IRInterfaceType*, IRComPtrType*> comPtrTypes;
+
+    void replaceTypeUses(IRInst* inst, IRInst* newValue)
+    {
+        List<IRUse*> uses;
+        for (auto use = inst->firstUse; use; use = use->nextUse)
+        {
+            uses.add(use);
+        }
+        for (auto use : uses)
+        {
+            switch (use->getUser()->getOp())
+            {
+            case kIROp_WitnessTableIDType:
+            case kIROp_WitnessTableType:
+            case kIROp_ThisType:
+            case kIROp_RTTIPointerType:
+            case kIROp_RTTIHandleType:
+            case kIROp_ComPtrType:
+                continue;
+            default:
+                break;
+            }
```

Comment:
This logic seems a little fiddly. It seems like we are getting burned by the fact that we use the `IRInterfaceType` directly to represent an existential, instead of having something like `IRExistentialType<Interface>` for that case, and then having an `IRInterface` not even *be* an `IRType`. That approach would let us be much more nuanced in a step like this, since really what you are trying to do here is replace *existentials* for a `[COM]` interface with the corresponding `ComPtr<...>` type.

(Ideally, at that point the AST->IR lowering logic would select between different subtypes of `IRExistentialType<...>`, with `ComPtr<...>` being just one case, so that we wouldn't even have to do the logic that is going on here at all.)